### PR TITLE
enhancement: add types for graph event names

### PR DESCRIPTION
**Pull request by cliffordfajardo**
## Description of change?
Added types for graph event names. Using `types` instead of `enums`  ([enum usage currently](https://github.com/antvis/G6/blob/master/packages/core/src/types/index.ts#L586))

Why `types` over `enum`s? 
- With types, user's of G6 do not need to import an enum into their file for type safety.
- `types` are "zero-cost", they do not produce JS code during build time/production. `enum`s produce JS code which makes the JS bundle bigger. More details in "References" section below.
- using enums would make it easier for users of typescript to read the documentation and copy/paste examples of using event handlers because the event name is just a string, whereas if we used enum's it gets a little more complicated

## Motivation for Change?
**1. While using `graph.on` , I am not getting autocompletion in my IDE for the event names**

https://user-images.githubusercontent.com/6743796/157084153-3a03eafb-6c29-48df-b866-5bf3035b2ab4.mp4

**2. Improved Development Experience**
   - Adding types to event names will allow text editors & IDE's to autocomplete event names as the user types & help them catch mistype errors. **This experience is not possible with `enum`s, as it would require the user to import the `enum`**.
   - Enhances developer experience, user's will likely not need to go to the documentation simply to find out all available graph event names.
   - Will help other `@antv` libraries such as `@antv/Graphin`  (or any other abstraction built on G6)

<br />

[Typescript code playground URL](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBDAnmYcByEAmwCSA7GYKAQwGMYBLCPAUQDdgC4BeAKDjgB84ByUgGwqkA1j3ZdemAEb8BQ0eO49MJAOYBnGMVhiOSlcVW6JytY0zH9ZgkUuS1-YMQZ3ThiAyiuVEMK4C2EACu6sCYEADueAHBoYyEXoq8gSHAQX5JPCmhgS6Z2cDBMDGpHrb5scCOzsCuYMR4wFo6mfV4ubWtDeauMMGkABaa2sWZfUGDHb39Az2ZpNSEIDD+jEE8ANysrKCQsAjIqDSYqrg2JORUtAxMbHq8ciLeMo8K924azaPvBkaZv3MftYEt4HE48kD3J4SnFzjDgFMKqUgt8TAUyol3gVqhC0ZVwlF4elXAsbMtVnh1lsduBoPAkCg4ABhCD+KQQfAJMiUaj0eIsdBYM5cy68m4wLa7OkHRlMhp0YjqTlEblXPm3eaCJ7-F5at4mX6gwzDFqQ054Cz-YHlM04zpA3zw+I2vGpRFYypFeEY+F2+EE6JI0LEsYzE2opTjSZlaYTWYWkmLUArNabbYAegAVJn2Jm4AABbDqUhQChgHl4XPM+WKuDKqBOUV4ODAcXqXNV-OgYj+MCOKs0PCqQTqAYALjgAxgMDA6jH6fTqgAbAA6BowOgrugUdRXdOMdPhUjqdPEMAUdPqmBVpkDCiNUITqczucL5drgib7e76jpgBeAyHhAx6nuel7irm6Y0ns9KHHAADiJBgAMAAqFD+PeqhXuoLCZFIwAAGbQEQ5gukoxAEQkDYWmRvD4URDbEJgmAUIQ-iuBRXLMaxCKuPRxENh0PHsZknFEIJZTCXxhHEekmDEIQUmiZRRByQpwBKe8-ENsJ34UFIFCCEggwNKcHEqaWbF6QZRmICZQ72iY2kaWxwyEPZZnKQkwlucAHmOUoznCQ2BENqO5neWxIVhQM0kMS5CK+cW1SYiYYmWYlWiEMlTipYFMkSdg-kRUQgRFQMpkBXRBVVMQiBel5RD8HVDVaTVqhIQM0XAKOkC7hWJVQB1Z5dYRMV9axVxxcRw3Id14WNUNnXzbFeE1Q06HqYNG3+Fta3xSWTiEGEnnvOlh3qSdVU8M5Cz8M1s7ALsDSYAsbIQINd0PXE4AvW97KuLNQxZX5FUOcYC5wChADyAAi0MTrtiD4XAESoI0YRwMQcAAAbOaqCoJLVFbqDjcAXYQmBwPecAESiQQNnAAAKABKAD8i0E+pDbNSTrh4EKoSOOQt6Va4wlfWeoSWu8fQQPwlBgMVmQRLMVR-hArKuNuwARDBytmgL2CAiYwwiAbcCQzD8MTgAgozADezkADTpQAvvU97wNgYANqQ6mYGzUOzIgWOMwL8A00e6jO3AUgoggIeoyR1PNjAswIUuV4tpS-grvtxGewQg1F8U1IMqgACyEA2cKKpNtndwmFoGTvOeeCDOGdT3u6JhtF3XTtDGg9zNB0oV4K2BXihcHMLjRvAGOAAkDsYNg9aqmK8Ruzjkq0vsE-HKc0+z7jV3Lw7R91xcFZXjve8wTKqAsu9J+MnPOP-RAF8v+yG8N+Ke+Y8D5wTlHgBU6g36oA-v7cBiof6sj-ucTe1xt672AbBRkiERpoQwkOKBApsHIVwZhbCD9x5wWrrXf+t9xQz3fnAKhhlr4oKvNSVgWYcxwDzIWHqJYywVirEQgYqcRQVhbG2KsnZuy9n7Nwyc05ZzzkXKudcX4dx7gPNHUCF4rwAGIv4AFovb1wrIY1s8RIIYKfghTqV40A9mgZPYABDuBX1ccyRBEAPFgIgTQtU4oJDCJIfguhcFuBMMcP4reBB6HAGpKSTQLYQA9j7MAAAjAAfQVIITAY5hH2McQKHgC8xyvDEIk+AMi0kACZsnEFyfkux4oHGrGKV-MpeoKnUCSdUxwABmepjSCktKKXPG661uJsW6XgXpKTZHAAACyZPvDkigeSRnxFaU4ngJALRa1YEAA) 

https://user-images.githubusercontent.com/6743796/157082266-d25b179c-f740-4ec4-81a1-87156ea0c59c.mp4

   
   
   
##### Checklist
- [ ] `npm test` passes
- [ ] tests and/or benchmarks are included
- [ ] commit message follows commit guidelines

### References
- https://fettblog.eu/tidy-typescript-avoid-enums/
- https://thisthat.dev/literal-union-type-vs-string-enums/
- [Example1: types do not produce Javascript code](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBDAnmYcByEAmwCSA7GYKAQwGMYBLCPAUQDdgC4BeAKDjgB84ByUgGwqkA1j3ZdemAEb8BQ0eO49MJAOYBnGMVhiOSlcVW6JytY0zH9ZgkUuS1-YMQZ3ThiAyiuVEMK4C2EACu6sCYEADueAHBoYyEXoq8gSHAQX5JPCmhgS6Z2cDBMDGpHrb5scCOzsCuYMR4wFo6mfV4ubWtDeauMMGkABaa2sWZfUGDHb39Az2ZpNSEIDD+jEE8ANysrKCQsAjIqDSYqrg2JORUtAxMbHq8ciLeMo8K924azaPvBkaZv3MftYEt4HE48kD3J4SnFzjDgFMKqUgt8TAUyol3gVqhC0ZVwlF4elXAsbMtVnh1lsduBoPAkCg4ABhCD+KQQfAJMiUaj0eIsdBYM5cy68m4wLa7OkHRlMhp0YjqTlEblXPm3eaCJ7-F5at4mX6gwzDFqQ054Cz-YHlM04zpA3zw+I2vGpRFYypFeEY+F2+EE6JI0LEsYzE2opTjSZlaYTWYWkmLUArNabbYAegAVJn2Jm4AABbDqUhQChgHl4XPM+WKuDKqBOUV4ODAcXqXNV-OgYj+MCOKs0PCqQTqAYALjgAxgMDA6jH6fTqgAbAA6BowOgrugUdRXdOMdPhUjqdPEMAUdPqmBVpkDCiNUITqczucL5drgib7e76jpgBeAyHhAx6nuel7irm6Y0ns9KHHAADiJBgAMAAqFD+PeqhXuoLCZFIwAAGbQEQ5gukoxAEQkDYWmRvD4URDbEJgmAUIQ-iuBRXLMaxCKuPRxENh0PHsZknFEIJZTCXxhHEekmDEIQUmiZRRByQpwBKe8-ENsJ34UFIFCCEggwNKcHEqaWbF6QZRmICZQ72iY2kaWxwyEPZZnKQkwlucAHmOUoznCQ2BENqO5neWxIVhQM0kMS5CK+cW1SYiYYmWYlWiEMlTipYFMkSdg-kRUQgRFQMpkBXRBVVMQiBel5RD8HVDVaTVqhIQM0XAKOkC7hWJVQB1Z5dYRMV9axVxxcRw3Id14WNUNnXzbFeE1Q06HqYNG3+Fta3xSWTiEGEnnvOlh3qSdVU8M5Cz8M1s7ALsDSYAsbIQINd0PXE4AvW97KuLNQxZX5FUOcYC5wChADyAAi0MTrtiD4XAESoI0YRwMQcAAAbOaqCoJLVFbqDjcAXYQmBwPecAESiQQNnAAAKABKAD8i0E+pDbNSTrh4EKoSOOQt6Va4wlfWeoSWu8fQQPwlBgMVmQRLMVR-hArKuNuwARDBytmgL2CAiYwwiAbcCQzD8MTgAgozADezkADTpQAvvU97wNgYANqQ6mYGzUOzIgWOMwL8A00e6jO3AUgoggIeoyR1PNjAswIUuV4tpS-grvtxGewQg1F8U1IMqgACyEA2cKKpNtndwmFoGTvOeeCDOGdT3u6JhtF3XTtDGg9zNB0oV4K2BXihcHMLjRvAGOAAkDsYNg9aqmK8Ruzjkq0vsE-HKc0+z7jV3Lw7R91xcFZXjve8wTKqAsu9J+MnPOP-RAF8v+yG8N+Ke+Y8D5wTlHgBU6g36oA-v7cBiof6sj-ucTe1xt672AbBRkiERpoQwkOKBApsHIVwZhbCD9x5wWrrXf+t9xQz3fnAKhhlr4oKvNSVgWYcxwDzIWHqJYywVirEQgYqcRQVhbG2KsnZuy9n7Nwyc05ZzzkXKudcX4dx7gPNHUCF4rwAGIv4AFovb1wrIY1s8RIIYKfghTqV40A9mgZPYABDuBX1ccyRBEAPFgIgTQtU4oJDCJIfguhcFuBMMcP4reBB6HAGpKSTQLYQA9j7MAAAjAAfQVIITAY5hH2McQKHgC8xyvDEIk+AMi0kACZsnEFyfkux4oHGrGKV-MpeoKnUCSdUxwABmepjSCktKKXPG661uJsW6XgXpKTZHAAACyZPvDkigeSRnxFaU4ngJALRa1YEAA)
- [Example2: Enums produce javascript code](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBMAdgVwLZwOIDYCiA3JeAbwCg44B6CuAYwlVQkQQMRgGcy4BhAGQEluAaTgBeOAHIaAGwCWNANYSANFwAiAIV59BI8RIAmAI2kz5S1eQCyAeQCqAZRxqbAdQByYyY2TtgBiAB3RBUuWzs1HDsABS8JHz9kMFDybht3ABUcAA0Mqxx3Ozi6NlAYVCRkFLhwpwKsgCU4hOBCYChq2px7DOaIX2B+mE77JxsANRwm-RaIAg7LGtGcW0m+gcYCEcccXhwAQTWZ-r9pYABDLcW1Bv2MBwz9ht79AyhzgHN2GHPYapu7gU1HE3p8kAZ-rcMCD3h9IYDMlMYWDSgt1FC9occMiPmdLsB4RgJkjXrC5u1CT0cUN-jcbLFSRBkoshDgAJoxOIKYAATyS1VZbJcHi5vICwWqrgAEjhdnFAgALVrSaoAMRs3EccQAZhAaL5qlo7NNJCZkGiuFQ4DB+jQFSxCJxyBl7NwpQ8ni9JDbkHbvr9hosXXY3atsfofXbNgSg66Ze5gRHbUrEBDFtx9u5xvsHAB9YNuj3PYrnRB4c7sABckYV-r+6cz2bzBalYZLZYr1eT0eqGazOfzcaB7fLVZr4Oq7hskUHIfdj2L+kQEAMwC7vtrP3rXCnM5bbaXK7XNZ7i13OFnbuHh9X67tE-TNisGhsl-nnuK9CMEDvm4DvafF83wPSQ6FQb9f1PLg0mfV8W2vUCvx-cdU2qZwMAvFsiy9CR-A+Y9kzrQMuHQzC4xA3CDHwyDyTQtQMLfBDKOolC0xIS1qGXVcHTYJ04HPXM0kRXJ8kKOIuLXEoYDKCoUEnacLx0YRxKPSszEUeSZ00bQBGUm812MUw5A0s8FNzLphU8fTKxacUQlMmcuk5ayWn5ByLy6eoSUkCSbJOVpUU0jzlkxI4fNUlo8SuHczK6CjfJaKDyAEroqRc-yaXc8zlmJE0JASjL5iC3M6QZcLbzeJlioBIlJjy3zQQ+ck0WSsyaq8+rVMatoWv4tqMQOML8q62EopjGKtKhbCVIq2EiOqqEZoM2EFoRRNyuWlE2Navd9jK4bbx+ZkJovaJM2m6ywFLeasrO9x4tUq7ECSvqZzupjfKeh92PIK0wO-ZQEBAaTU3YOBtSgeg4AkviYKAoSshEgoin0f6fykmTKgA2DBN0vREPA9HjIsaDANfbSlPxqQkMrQz1JJ1IyeynZLM-Qm-IGOzsaApz9rRjnEmOxmcc8xE8v5loeu518ulC8MCYgyKLmi4WeeWCiJf8l64ZlnK7BwzWBky0mRZyuq2cVwqKUfHHSotn9KqFngmZq3L7dpskiptoD2rF93usC73yYGrF-dG5XxtV4O7guhWHbmrdiKjkrFtRmnGullO1rDrbM8ePmaaOzO7tj6n2aem6TaAu6NZpp7taZ96E3dr7UNUDiECo4AeI4Ej6MU9JEbyZG4jwySmGk4HZKqRZSNx3RR67tTibohiKbxxfqLplfZ-75mnFZ-Qx4F-wgnsvuGN5ze11cp259Fxpr5PqXd8vkLBvl5ib-8sbV+CnYKLH0SrRV+-8xj6yfrMZASc4D3zNt5L+J9mp-xTvSJ+jsUGu3NkfJejVkGgKzhhP2OCt6whfhfC8NU5boPDviUIHcaCllHD3WGjYBwIxyMPMSqMmGdgxlPLGDZ+x5kpiOPhO9oJsLzOvBePCOxVm3uYXsUj97ODcFZUCvCqy2TPso4RqjnKaPkSfNykj9EPwQYw4xktA5mKbKo6hcjRwn1-kI+xcViRiO0VrEBdiBypQgU4zsUCYF9ncfA8WWikFez8dIhoaCgkKMhk7MJA5faP0SR7FE0leqpLiXcRxRjnGNVcbEwhpcrHFITv+NxaTU5FM7BnWp+SiHrSkFEgO20eAqPzl46s5wUkqJLguA2USK6Jz0fYmunjMn118akIZmYmKVM7K3NiHcYCyFQLIRAHwWFcA0DgdUDQphAgQUYYAuooDtHBNbLg+xVSNBOQmBB5xtQ5OuamO55BDnHIOGoNQ-AshWDiBcq5FwDAGFkNJVA1QHmNH2ACoFKw4hvJyecSF0LgCwsWL8mwJyTlhmRSC-QYLoA3OjFinF9zHlTEJcSYlqL3ntGuZSmFhojn4qiNENQ+wsiMtJZc8lSQDDnGklSuFtKGgxF5fy4FTKckirFcACVuLOUnOJeMfgDh+AaH4AIDIbI3SZgwqCoV1yqV4FkOwWQRhZByBgDyO0pZ8KSsaJq7Vur9VAqNVKE1n80XtEtda219roVOoVC6yOcA8UauBR6LIxr3CmsFeCql-ppLOt2dG+FUxiUJpwEmlNkhA1QHTT8TNkbs0cr+cSk5qoTkOClGatNMLrkQ2AOwBUbq83AvrY25t+hS1UvbdcrtNauX5seFkBwfADh5TJRamFGbO0yAuL1XNDQp18pwLOzEeVh3LorauvEvVY0rAUkWz+i7sVHiza6xYm7bCRCvQq9ojBVz3ujee3g+w2RpVNOa4A0hzg8mNuQTdv7-2BJLcyqAIGwPQInSc7gJyd2kRbeSmg1zlVjx7Q0VDBwsgYaHXB7DFxpJ4bVX8jAtxohSn7buqU0QbA6oyPwdImHrkfHeGABUo7O0KkgDazZTB8O0b2gxo5A6WNsY4xoiQpaeMDP45csdQmIAidkGJ6jXKJP0cY02rjwBlN8YE+Ox9Ur9NSYbUxt9UBTOqY7RZg56qDjuH4FYHdxnSxbOVfhzMnnvOkfRYgPz0lkM4DOvwTIxmro7JgZu6LsWQvtHi2wSLaReC-uiHUbId01A62M3QaQIGwB+FAE9AwaN8NZZy3lgrRXUtQBK2Vir4BSzVaQvQ8g1mC2vv0I5ldX6AvcHY9mRouw+XyYcPZ84NBNnlhycBsV2nEDsCCk4PYY2BsbVOMABbI3Yw2F4Ox6Iu2JA2ggNITZYAjtcGlLKXgAAtGwT55RKmAwALwgPQaoWqcCuBY88C7VrgCBEgLAe75AarniYo1CS31yAemEBd-0ih7sdysBAO1Zwe5wHYEkSHMAuC9IjAM6o0QYuFhGXEMAOy-QTMWFT9woYZmSHp4gKM8y4DDI-PocZNSuDTKGnMlWvOlnNwF6Wb6ABfIAA)

### CODE REVIEW

**Comment by cliffordfajardo** [comment]
@Yanyan-Wang 
I think this would be a big improvement in the G6 developer experience.
It would require changing a few places in the codebase where the [`G6Event` enum is used](https://github.com/antvis/G6/blob/master/packages/core/src/types/index.ts#L586)

File: packages/core/src/types/index.ts [hunk 1]
```diff
@@ -12,6 +12,149 @@ export type UpdateType = 'move' | 'bbox' | 'style' | 'bbox|label' | 'style|label
...
+  | 'dragover'
+  | 'mouseenter'
+  | 'mousemove'
+  | 'mouseout'
+  | 'mouseover'
+  | 'mouseleave'
+  | 'mousedown'
+  | 'mouseup'
+  | 'contextmenu';
+
```
Comment (cliffordfajardo) [inline packages/core/src/types/index.ts:57]: @Yanyan-Wang 
I chose to explicitly list the event name for each type (node, edge, combo etc) to make the code maintenance a bit easier
- if a developer wants to know all event names for node/graph/combo, they can quickly go directly to the list
- adding new types might be easier mentally, the person only needs to go directly to the entity they are interested in

The list when updated will automatically update this type below
```ts
// template literal types: https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html
export type NodeEventType = `node:${NodeInteractionEvent}`;
export type EdgeEventType = `edge:${EdgeInteractionEvent}`;
export type ComboEventType = `combo:${ComboInteractionEvent}`;
export type CanvasEventType = `canvas:${ComboInteractionEvent}`;
export type GraphTimingEventType = GraphTimingEvents;
export type MobileInteractionEventType = MobileInteractionEvent;
export type G6EventName = NodeEventType | EdgeEventType | ComboEventType | CanvasInteractionEvent | GraphTimingEventType | MobileInteractionEventType;
```

File: packages/core/src/types/index.ts [hunk 2]
```diff
@@ -12,6 +12,149 @@ export type UpdateType = 'move' | 'bbox' | 'style' | 'bbox|label' | 'style|label
...
+  | 'dragleave'
+  | 'dragover'
+  | 'mouseenter'
+  | 'mousemove'
+  | 'mouseout'
+  | 'mouseover'
+  | 'mouseleave'
+  | 'mousedown'
+  | 'mouseup'
+  | 'contextmenu';
```
Comment (cliffordfajardo) [inline packages/core/src/types/index.ts:56]: @Yanyan-Wang 
I chose to explicitly list the event name for each type (node, edge, combo etc) to make the code maintenance a bit easier
- if a developer wants to know all event names for node/graph/combo, they can quickly go directly to the list
- adding new types might be easier mentally, the person only needs to go directly to the entity they are interested in

The list when updated will automatically update this type below
```ts
// template literal types: https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html
export type NodeEventType = `node:${NodeInteractionEvent}`;
export type EdgeEventType = `edge:${EdgeInteractionEvent}`;
export type ComboEventType = `combo:${ComboInteractionEvent}`;
export type CanvasEventType = `canvas:${ComboInteractionEvent}`;
export type GraphTimingEventType = GraphTimingEvents;
export type MobileInteractionEventType = MobileInteractionEvent;
export type G6EventName = NodeEventType | EdgeEventType | ComboEventType | CanvasInteractionEvent | GraphTimingEventType | MobileInteractionEventType;
```

**Review by cliffordfajardo — COMMENTED** [review]
.

**Comment by cliffordfajardo** [comment]
@Yanyan-Wang 
I don't have much context on the history of the codebase so I have a few questions 😄 
I noticed that the `core/tests/unit/implement-graph.ts` seems to be mocking the behavior of the real G6 library library?
- Why is the mocking required?
- Is it possible to test against the real implementation?

My first thought is that this mocking probably is concerned with mocking the canvas or some other technology?
Currently I'm working on a new github pull request for Graphin that involves adding more unit tests.
One of the issues I initially ran into was getting errors because Jest does not mock the canvas by default. I was able to resolve this though using `canvas` package
- https://github.com/antvis/Graphin/compare/master...cliffordfajardo:cliffordfajardo/add-tests-to-graphin-component?expand=1#diff-7b20da53cef72b0d99b442453618c091c4ed8e8b98c797ffb288bceff63d72b4R31https://github.com/antvis/Graphin/compare/master...cliffordfajardo:cliffordfajardo/add-tests-to-graphin-component?expand=1#diff-7b20da53cef72b0d99b442453618c091c4ed8e8b98c797ffb288bceff63d72b4R31

**Comment by cliffordfajardo** [comment]
After running the `npm run build:all` command I was getting these erros from  `src/behavior/drag-canvas.ts` file

```
Property 'getEdges' does not exist on type 'IGraph'.ts(2339)
Property 'getNodes' does not exist on type 'IGraph'.ts(2339)
```
CleanShot 2022-03-07 at 12 40 10@2x

**Comment by cliffordfajardo** [comment]
3. The IG6GraphEvent is missing lots of types
It seems like lots of the methods it has are based on the DOM event's object, but a modified version?
I think we may be able to type it and extract out properties we don't need?

```
type IG6GraphEvent = Omit<MouseEvent, 'prop1' | 'prop2' ....>
```

Which Github repository does `GraphEvent` class live in? Could not find the exact file on github.com to understand the code

CleanShot 2022-03-07 at 13 18 04@2x

File: packages/core/src/types/index.ts [hunk 3]
```diff
@@ -12,6 +12,188 @@ export type UpdateType = 'move' | 'bbox' | 'style' | 'bbox|label' | 'style|label
...
+  // TODO: maybe we need a `beforeactivaterelations` created in future PR?
+  | 'afteractivaterelations'
+  | 'nodeselectChange'
+  | 'itemcollapsed'
+  | 'tooltipchange'
+  | 'wheelzoom'
+  | 'viewportchange'
+  | 'dragnodeend'
+  | 'stackchange'
+  // TODO: Are {before,after}paint deprecated? They are not in docs, but they were in the G6Event enum.
```
Comment (Yanyan-Wang) [inline packages/core/src/types/index.ts:169]: Yes they are deprecated but exist in the code for compatibility

File: packages/core/src/types/index.ts [hunk 4]
```diff
@@ -12,6 +12,188 @@ export type UpdateType = 'move' | 'bbox' | 'style' | 'bbox|label' | 'style|label
...
+  | 'beforegraphrefresh'
+  | 'aftergraphrefresh'
+  | 'beforeanimate'
+  | 'afteranimate'
+  | 'beforecreateedge'
+  | 'aftercreateedge'
+  | 'beforecollapseexpandcombo'
+  | 'aftercollapseexpandcombo'
+  | 'graphstatechange'
+  // TODO: maybe we need a `beforeactivaterelations` created in future PR?
```
Comment (Yanyan-Wang) [inline packages/core/src/types/index.ts:160]: Of course if there are related requirements

File: packages/core/src/types/index.ts [hunk 5]
```diff
@@ -12,6 +12,188 @@ export type UpdateType = 'move' | 'bbox' | 'style' | 'bbox|label' | 'style|label
...
+  | 'mouseover'
+  | 'mouseleave'
+  | 'panestart'
+  | 'panmove'
+  | 'panend'
+  | 'touchstart'
+  | 'touchmove'
+  | 'touchend'
+  | 'wheel'
+  | 'contextmenu';
```
Comment (Yanyan-Wang) [inline packages/core/src/types/index.ts:69]: There are no 'keyup', 'keydown', and 'wheel' for NodeInteractionEvent / EdgeInteractionEvent / ComboInteractionEvent

File: packages/core/src/types/index.ts [hunk 6]
```diff
@@ -12,6 +12,188 @@ export type UpdateType = 'move' | 'bbox' | 'style' | 'bbox|label' | 'style|label
...
+  | 'mousedown'
+  | 'mouseup'
+  | 'touchstart'
+  | 'touchmove'
+  | 'touchend'
+  | 'wheel'
+  | 'contextmenu';
+
+export type ComboInteractionEvent = NodeInteractionEvent;
+export type CanvasInteractionEvent =
```
Comment (Yanyan-Wang) [inline packages/core/src/types/index.ts:96]: looks like something is missed, e.g. 'tap'. Maybe it should correspond to the old G6Event

File: packages/core/src/types/index.ts [hunk 7]
```diff
@@ -24,6 +206,8 @@ export interface IG6GraphEvent extends GraphEvent {
   detail: number;
   key?: string;
   target: IShapeBase & ICanvas;
+  //type:
+  //name: should we add types to `@antv/g-base/event`?
```
Comment (Yanyan-Wang) [inline packages/core/src/types/index.ts:210]: How about extends it here?

**Comment by Yanyan-Wang** [comment]
GraphEvent is in https://github.com/antvis/g/tree/master/packages

but @antv/g is in progress for a new version witch is totally different from the old one in the branch 'next'

**Comment by Yanyan-Wang** [comment]
> imagine I am a contributor or developer using G6 and I use my code editor (VS Code) to jump to the definition of a Node's event's, I can see everything in one place, without needing to reference other types (like a common event type).

That is great!

**Comment by Yanyan-Wang** [comment]
> I noticed that the [core/tests/unit/implement-graph.ts file](https://github.com/antvis/G6/blob/master/packages/core/tests/unit/implement-graph.ts) seems to be mocking the behavior of the real G6 library library?

Yes, it is mocking for test specs in the g6-core package. If you want to test the complete G6, maybe try the test specs in packages/pc and link the packages/core


> After running the npm run build:all command I was getting these erros from src/behavior/drag-canvas.ts file

getNodes and getEdges are defined here (IAbstractGraph): https://github.com/antvis/G6/blob/5dd6a79572fcdd79128f1fb0bc4bf8f69fd932e7/packages/core/src/interface/graph.ts#L303

and the [IGraph](https://github.com/antvis/G6/blob/5dd6a79572fcdd79128f1fb0bc4bf8f69fd932e7/packages/pc/src/interface/graph.ts#L8) is extended from the IAbstractGraph

**Comment by cliffordfajardo** [comment]
> GraphEvent is in https://github.com/antvis/g/tree/master/packages
> but @antv/g is in progress for a new version witch is totally different from the old one in the branch 'next'

@Yanyan-Wang - do you know how this will effect G6 and if so when the `@antv/g` next branch code might be integrated into G6 (is it for G6 version 5.0)?

I imagine these are internal upgrades to the rendering technology or will there be significant API changes?
I read the docs and I think it direction the `@antv/g6` team is going in is awesome 🎉 
- Specification: https://g-next.antv.vision/api.html

File: packages/core/src/types/index.ts (hunk 6 above)
Comment (Yanyan-Wang) [reply packages/core/src/types/index.ts:90]: Ahh, sorry I misunderstood it. That is right, tap should be in the MobileInteractionEvent

**Comment by Yanyan-Wang** [comment]
> > GraphEvent is in https://github.com/antvis/g/tree/master/packages
> > but @antv/g is in progress for a new version witch is totally different from the old one in the branch 'next'
> 
> @Yanyan-Wang - do you know how this will effect G6 and if so when the `@antv/g` next branch code might be integrated into G6 (is it for G6 version 5.0)?
> 
> I imagine these are internal upgrades to the rendering technology or will there be significant API changes? I read the docs and I think it direction the `@antv/g6` team is going in is awesome 🎉
> 
> * Specification: https://g-next.antv.vision/api.html

Next @antv/g supports WebGL rendering and computing. And the canvas and WebGL are convertible in the runtime. That will help G6 to handle large graph visualization. The upgrading will be complete this year (G6 v4.x).

We are trying to build up a g6-g-adapter to eliminate the breaking-changes for users. 

And another reason for implementing the g6-g-adapter is, now, when user custom a node or an edge, they have to learn the @antv/g rendering api. In the future, we want to design a more 'graph-vis-grammar' for graph rendering and node/edge customization for users instead of the basic graphics apis from rendering engine.

File: packages/core/src/types/index.ts [hunk 8]
```diff
@@ -12,6 +12,188 @@ export type UpdateType = 'move' | 'bbox' | 'style' | 'bbox|label' | 'style|label
...
+export type CommonInteractionEvent =
+  | 'click'
+  | 'dblclick'
+  | 'dragstart'
+  | 'drag'
+  | 'dragend'
+  | 'dragenter'
+  | 'dragleave'
+  | 'dragover'
+  | 'dragout' // TODO: not in documentation page: https://g6.antv.vision/en/docs/api/Event/
```
[the remark this answers is not in the document]
Comment (Yanyan-Wang) [reply packages/core/src/types/index.ts:24]: I'm not really clearly remember the difference. It seems that the `dragout` is bubbling and the `dragleave` is not

**Comment by Yanyan-Wang** [comment]
v4.6.4 published

**Pull request merged**